### DISCUSSION TITLE
Best way to handle short password expiry in connection pool

### DISCUSSION THREAD

**Discussion by bluetech** [psycopg/psycopg#500] [Q&A]
Some PG scenarios use short-lived passwords, that are issued dynamically and expire after a short period of time. Examples are [AWS RDS IAM authentication](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html) (the one I'm familiar with) and [Vault](https://developer.hashicorp.com/vault/docs/secrets/databases/postgresql). After expiry (15 minutes in RDS), existing connections don't disconnect or anything, but new connections require a new API call to get a fresh password.

What is the best way to handle this with `ConnectionPool`? Some solutions I thought of:

1. Subclass `ConnectionPool` to make `conninfo`/`kwargs` dynamic, getting a fresh password if it expired (or about to expire).
2. Add a periodic maintenance task which actively refreshes the password.
3. Somehow use the `reconnect_failed` callback.

Ideally there won't be any added latency of a getconn waiting for a refresh, so option 2 seems good.

**Comment by dvarrazzo** [comment]
I think it would be solved well with a background task in your process (not in the pool) that would periodically fetch a new password and set [the `PGPASSWORD` env var](https://www.postgresql.org/docs/current/libpq-envars.html) in the process. The password would be used automatically by the libpq at the moment a new connection is requested.

You could even write an external script requesting periodically a new password and writing it in the `.pgpass` file. I think this could potentially have race conditions (I don't know how atomic is to open and write the file), but would need no change in a program.

**Reply by bluetech** [reply]
Thanks for the answer.

An envvar is not something I'd want to use for this - it's process global, leaks into subprocesses (and often various logs/dumps) and [not thread safe](https://github.com/python/cpython/issues/83556). Also, wouldn't work in [multi-db scenarios](https://docs.djangoproject.com/en/4.1/topics/db/multi-db/) AFAIK.

A pgpass file is not applicable in my case because I run in containers with a readonly filesystem (security/compliance reasons...).

If I end up doing RDS IAM with psycopg-pool, I'll try some solutions and report back.

**Reply by dvarrazzo** [reply]
Thank you for the info about the non-thread-safety of setenv.

AFAICS, looking at the current implementation, the pool doesn't do anything unsavory with the `conninfo` and the `kwargs` params. At the moment, changing these attributes on the connection, and even changing inplace the kwargs passed to the constructor, should work as expected when a new connection is created.

This property is only accidental. If we want to rely on it, we should add regression tests. I would make sure that the following statements work as expected:

```python
pool.conninfo = new_conninfo
pool.kwargs = new_kwargs
pool.kwargs["param"] = new_param
```

whereas I would make no guarantee about whether the init kwargs is used as it is or it is copied, i.e. the following would have an undefined behaviour:

```python
kwargs = some_params
pool = ConnectionPool(..., kwargs=kwargs, ...)
kwargs["param"] = new_param
```

**Reply by bluetech** [reply]
Thanks, I'll do it once I get it up and running and tested on my side (*probably* around when Django 4.2 with psycopg support is released)

**Comment by juan-fonto** [comment]
Hi @bluetech, 
We are now facing the same issue with RDS IAM auth, did you manage to find a solution for this?
Thanks
